### PR TITLE
Add Trait `Floor` to migration built-in functions like `floor_*

### PR DESCRIPTION
**Pull request by iosmanthus**
Signed-off-by: Iosmanthus Teng <myosmanthustree@gmail.com>

###  What have you changed?
Add trait `floor` to migration built-in functions:
* `floor_real`
* `floor_dec_to_dec`
* `floor_dec_to_int`
* `floor_int_to_int`

###  What is the type of changes?
- New feature (a change which adds functionality)

###  How is the PR tested?
unit test suite.

### CODE REVIEW

File: components/tidb_query/src/rpn_expr/impl_math.rs [hunk 1]
```diff
@@ -112,4 +180,84 @@ mod tests {
...
+                .evaluate::<Int>(ScalarFuncSig::FloorDecToInt)
+                .unwrap();
+            assert_eq!(expected, output);
+        }
+    }
+
+    #[test]
+    fn test_floor_int_to_int() {
+        let cases = vec![
+            (1, 1),
```
Comment (mapleFU) [inline components/tidb_query/src/rpn_expr/impl_math.rs]: What about the numbers less than 0?

Comment (iosmanthus) [reply components/tidb_query/src/rpn_expr/impl_math.rs]: Thanks, I will add this case.

Comment (iosmanthus) [reply components/tidb_query/src/rpn_expr/impl_math.rs]: Wait! I have already add `std::i64::MIN` to test cases.

File: components/tidb_query/src/rpn_expr/impl_math.rs [hunk 2]
```diff
@@ -2,8 +2,76 @@ use tidb_query_codegen::rpn_fn;
...
+    type Output: Evaluable;
+    fn floor(_ctx: &mut EvalContext, arg: &Self::Input) -> Result<Option<Self::Output>>;
+}
+
+#[rpn_fn(capture = [ctx])]
+pub fn floor<T: Floor>(ctx: &mut EvalContext, arg: &Option<T::Input>) -> Result<Option<T::Output>> {
+    if let Some(arg) = arg {
+        T::floor(ctx, arg)
+    } else {
+        Ok(None)
```
Comment (mapleFU) [inline components/tidb_query/src/rpn_expr/impl_math.rs]: Could we cover the case for `OK(None)` here?

File: components/tidb_query/src/rpn_expr/impl_math.rs [hunk 3]
```diff
@@ -112,4 +180,85 @@ mod tests {
             assert_eq!(output, expect_output, "{:?}", arg);
         }
     }
+
+    #[test]
+    fn test_floor_real() {
+        let cases = vec![
+            (3.0, 3.5),
```
Comment (Hoverbear) [inline components/tidb_query/src/rpn_expr/impl_math.rs]: Hm, do we need to worry about floating point precision here?

For example:

```rust
fn main() {
    let maybe = 16140901064495871255;
    assert_eq!(maybe, (maybe as f64) as u64);
}
```

Yields:

```
   Compiling playground v0.0.1 (/playground)
    Finished dev [unoptimized + debuginfo] target(s) in 0.65s
     Running `target/debug/playground`
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `16140901064495871255`,
 right: `16140901064495872000`', src/main.rs:3:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

Comment (iosmanthus) [reply components/tidb_query/src/rpn_expr/impl_math.rs]: `floor` returns the largest integer less than or equal to a number (an `f64`). Its behavior is as same as https://doc.rust-lang.org/std/primitive.f64.html#method.floor. So we could ignore floating-point precision here.

Comment (Hoverbear) [reply components/tidb_query/src/rpn_expr/impl_math.rs]: Hm, ok. I did some investigation and it seems like REAL in SQL is also intended to be a float. :) Thanks for looking at this.

**Comment by breeswish** [comment]
Updates?

**Pull request merged**